packages/react-devtools-extensions/src/parseHookNames/loadSourceAndMetadata.js

```diff
@@ -428,47 +459,109 @@ function loadSourceFiles(
+            },
+          );
+        };
+      }
 
-    const fetchPromise =
-      dedupedFetchPromises.get(runtimeSourceURL) ||
-      fetchFileFunction(runtimeSourceURL).then(runtimeSourceCode => {
-        // TODO (named hooks) Re-think this; the main case where it matters is when there's no source-maps,
-        // because then we need to parse the full source file as an AST.
-        if (runtimeSourceCode.length > MAX_SOURCE_LENGTH) {
-          throw Error('Source code too large to parse');
-        }
+      const fetchPromise =
+        dedupedFetchPromises.get(runtimeSourceURL) ||
+        fetchFileFunction(runtimeSourceURL).then(runtimeSourceCode => {
+          // TODO (named hooks) Re-think this; the main case where it matters is when there's no source-maps,
+          // because then we need to parse the full source file as an AST.
+          if (runtimeSourceCode.length > MAX_SOURCE_LENGTH) {
+            throw Error('Source code too large to parse');
+          }
 
-        if (__DEBUG__) {
-          console.groupCollapsed(
-            `loadSourceFiles() runtimeSourceURL "${runtimeSourceURL}"`,
-          );
-          console.log(runtimeSourceCode);
-          console.groupEnd();
-        }
+          if (__DEBUG__) {
+            console.groupCollapsed(
+              `loadSourceFiles() runtimeSourceURL "${runtimeSourceURL}"`,
+            );
+            console.log(runtimeSourceCode);
+            console.groupEnd();
+          }
 
-        return runtimeSourceCode;
-      });
-    dedupedFetchPromises.set(runtimeSourceURL, fetchPromise);
+          return runtimeSourceCode;
+        });
+      dedupedFetchPromises.set(runtimeSourceURL, fetchPromise);
 
-    setterPromises.push(
-      fetchPromise.then(runtimeSourceCode => {
-        hookSourceAndMetadata.runtimeSourceCode = runtimeSourceCode;
-      }),
-    );
+      setterPromises.push(
+        fetchPromise.then(runtimeSourceCode => {
+          hookSourceAndMetadata.runtimeSourceCode = runtimeSourceCode;
+        }),
+      );
+    }
   });
 
   return Promise.all(setterPromises);
 }
+
+export function prefetchSourceFiles(
+  hooksTree: HooksTree,
+  fetchFileWithCaching: FetchFileWithCaching | null,
+): void {
+  // Deduplicate fetches, since there can be multiple location keys per source map.
+  const dedupedFetchPromises = new Set();
+
+  let fetchFileFunction = null;
+  if (fetchFileWithCaching != null) {
+    // If a helper function has been injected to fetch with caching,
+    // use it to fetch the (already loaded) source file.
+    fetchFileFunction = url => {
+      return withAsyncPerformanceMark(
+        `[pre] fetchFileWithCaching("${url}")`,
+        () => {
+          return ((fetchFileWithCaching: any): FetchFileWithCaching)(url);
+        },
+      );
+    };
+  } else {
+    fetchFileFunction = url => fetchFile(url, '[pre] fetchFile');
+  }
+
+  const hooksQueue = Array.from(hooksTree);
```

Comment:
are you doing this custom traversal to avoid flattening the list first and then fetching?

Reply:
Yup. Just voices recursion.